multishard_mutation_query.cc

```diff
@@ -560,21 +560,22 @@ future<> read_context::lookup_readers(db::timeout_clock::time_point timeout) noe
     }
     try {
         return _db.invoke_on_all([this, cmd = &_cmd, ranges = &_ranges, gs = global_schema_ptr(_schema),
-                gts = tracing::global_trace_state_ptr(_trace_state), timeout] (replica::database& db) mutable {
+                gts = tracing::global_trace_state_ptr(_trace_state), timeout] (replica::database& db) mutable -> future<> {
             auto schema = gs.get();
-            auto querier_opt = db.get_querier_cache().lookup_shard_mutation_querier(cmd->query_uuid, *schema, *ranges, cmd->slice, gts.get(), timeout);
             auto& table = db.find_column_family(schema);
+            auto querier_opt = db.get_querier_cache().lookup_shard_mutation_querier(cmd->query_uuid, *schema, *ranges, cmd->slice, gts.get(), timeout);
             auto& semaphore = this->semaphore();
             auto shard = this_shard_id();
 
             if (!querier_opt) {
                 _readers[shard] = reader_meta(reader_state::inexistent);
-                return;
+                co_return;
             }
 
             auto& q = *querier_opt;
 
             if (&q.permit().semaphore() != &semaphore) {
+                co_await q.close();
```

Comment:
`reader_meta` constructor may throw too as it's calling `std::make_unique` for the remote parts.
I think a more holistic approach would be to catch any exception after setting querier_opt and to `co_await q.close()` if the exception_ptr is caught.

Reply:
> `reader_meta` constructor may throw too as it's calling `std::make_unique` for the remote parts.
> 
> I think a more holistic approach would be to catch any exception after setting querier_opt and to `co_await q.close()` if the exception_ptr is caught.

By the time the reader_meta constructor runs, the reader is already registered as inactive, and will be closed by the semaphore. The reader is extracted from the querier asap, with two checks in between. One of these was a throwing check, but there is nothing else to put a try-catch around.

Reply:
> also, please extract the coroutinization of the lambda into a separate patch ...

I think the coroutinization changes are minimal but ok.

Reply:
> > `reader_meta` constructor may throw too as it's calling `std::make_unique` for the remote parts.
> > I think a more holistic approach would be to catch any exception after setting querier_opt and to `co_await q.close()` if the exception_ptr is caught.
> 
> By the time the reader_meta constructor runs, the reader is already registered as inactive, and will be closed by the semaphore. The reader is extracted from the querier asap, with two checks in between. One of these was a throwing check, but there is nothing else to put a try-catch around.

Ok, fair enough, but let's add a comment explaining this for future selves